clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -121,29 +375,174 @@ mlir::Value CIRGenFunction::emitScalarExpr(const Expr *e) {
+  case CK_NoOp: {
+    auto v = Visit(const_cast<Expr *>(subExpr));
+    if (v) {
+      // CK_NoOp can model a pointer qualification conversion, which can remove
+      // an array bound and change the IR type.
+      // FIXME: Once pointee types are removed from IR, remove this.
+      mlir::Type t = cgf.convertType(destTy);
+      if (t != v.getType())
+        cgf.getCIRGenModule().errorNYI("pointer qualification conversion");
+    }
+    return v;
+  }
+
+  case CK_NullToPointer: {
+    if (MustVisitNullValue(subExpr))
+      cgf.getCIRGenModule().errorNYI(
+          subExpr->getSourceRange(),
+          "ignored expression on null to pointer cast");
+
+    // Note that DestTy is used as the MLIR type instead of a custom
+    // nullptr type.
+    mlir::Type ty = cgf.convertType(destTy);
+    return builder.getNullPtr(ty, cgf.getLoc(subExpr->getExprLoc()));
+  }
+
   case CK_LValueToRValue:
-    assert(cgf.getContext().hasSameUnqualifiedType(e->getType(), destTy));
-    assert(e->isGLValue() && "lvalue-to-rvalue applied to r-value!");
-    return Visit(const_cast<Expr *>(e));
+    assert(cgf.getContext().hasSameUnqualifiedType(subExpr->getType(), destTy));
+    assert(subExpr->isGLValue() && "lvalue-to-rvalue applied to r-value!");
+    return Visit(const_cast<Expr *>(subExpr));
 
   case CK_IntegralCast: {
-    assert(!cir::MissingFeatures::scalarConversionOpts());
-    return emitScalarConversion(Visit(e), e->getType(), destTy,
+    ScalarConversionOpts opts;
+    if (auto *ice = dyn_cast<ImplicitCastExpr>(ce)) {
+      if (!ice->isPartOfExplicitCast())
+        opts = ScalarConversionOpts(cgf.sanOpts);
+    }
+    return emitScalarConversion(Visit(subExpr), subExpr->getType(), destTy,
+                                ce->getExprLoc(), opts);
+  }
+
+  case CK_FloatingRealToComplex:
+  case CK_FloatingComplexCast:
+  case CK_IntegralRealToComplex:
+  case CK_IntegralComplexCast:
+  case CK_IntegralComplexToFloatingComplex:
+  case CK_FloatingComplexToIntegralComplex:
+    llvm_unreachable("scalar cast to non-scalar value");
+
+  case CK_PointerToIntegral: {
+    assert(!destTy->isBooleanType() && "bool should use PointerToBool");
+    if (cgf.cgm.getCodeGenOpts().StrictVTablePointers)
+      cgf.getCIRGenModule().errorNYI(subExpr->getSourceRange(),
+                                     "strict vtable pointers");
+    return builder.createPtrToInt(Visit(subExpr), cgf.convertType(destTy));
+  }
+  case CK_ToVoid:
+    cgf.getCIRGenModule().errorNYI(subExpr->getSourceRange(),
+                                   "ignored expression on void cast");
+    return nullptr;
+
+  case CK_IntegralToFloating:
+  case CK_FloatingToIntegral:
+  case CK_FloatingCast:
+  case CK_FixedPointToFloating:
+  case CK_FloatingToFixedPoint: {
+    if (kind == CK_FixedPointToFloating || kind == CK_FloatingToFixedPoint) {
+      cgf.getCIRGenModule().errorNYI(subExpr->getSourceRange(),
+                                     "fixed point casts");
+      return {};
```

Comment:
what is the difference between `return nullptr` and `return {}` for this type?

Reply:
There should be no difference. mlir::Value is a wrapper around an implementation pointer, so returning `nullptr` will construct an mlir::Value with a null implementation, and the default constructor does the same. Of course, that could change. I tried to standardize on `{}` in the constant emitter.

Reply:
Good to know you standardized on `{}`. I changed all instances of `return nullptr` to that.